Lib/functools.py

```diff
@@ -883,6 +883,7 @@ def wrapper(*args, **kw):
     wrapper.dispatch = dispatch
     wrapper.registry = types.MappingProxyType(registry)
     wrapper._clear_cache = dispatch_cache.clear
+    wrapper._funcname = funcname
```

Comment:
Why not define ``__set_name__`` in ``singledispatchmethod``, rather than altering the code in ``singledispatch``? It seems to be a ``singledispatchmethod``-specific problem, so I think that would be better.

In the ``singledispatchmethod`` class, you could define ``__set_name__`` like this:

```
def __set_name__(self, owner, name):
    self._func_name = name
```

and then when raising the ``TypeError``, change the f-string to ``{self._func_name}`` rather than ``{self.dispatcher._funcname}``.

Other than that, LGTM 🙂

Docs for ``__set_name__`` are here: https://docs.python.org/3/howto/descriptor.html#customized-names

Reply:
It is perhaps even overkill, since we can get the name from `self.func`.